File: _Linked List/cpp/23_Merge_k_Sorted_Lists.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <iterator>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

ListNode* vector2ListNode(vector<int> *ar_ptr){
    ListNode* list = new ListNode(0);
    ListNode* ptr = list;
    for(vector<int>::iterator i=ar_ptr->begin();i<ar_ptr->end();i++){
        ListNode* newNode = new ListNode(*i);
        ptr->next = newNode;
        ptr = ptr->next;
    }
    return list->next;
}

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if(lists.empty()){
            return nullptr;
        }
        while(lists.size()>1){
            lists.push_back(mergeTwoLists(lists[0],lists[1]));
            lists.erase(lists.begin());
            lists.erase(lists.begin());
        }
        return lists.front();
    }

    ListNode *mergeTwoLists(ListNode *l1, ListNode *l2) {
        if(l1 == nullptr){
            return l2;
        }
        if(l2 == nullptr){
            return l1;
        }
        if(l1->val <= l2->val){
            l1->next = mergeTwoLists(l1->next, l2);
            return l1;
        }
        else{
            l2->next = mergeTwoLists(l1, l2->next);
            return l2;
        }
    }
};

int main(){
    vector<int> a = {45, 47, 49, 80, 97};
    vector<int> b = {20, 24, 31, 65, 87};
    vector<int> c = {33, 45, 56, 64, 93};
    vector<int> d = { 3, 19, 35, 38, 59};

    ListNode* l1 = vector2ListNode(&a);
    ListNode* l2 = vector2ListNode(&b);
    ListNode* l3 = vector2ListNode(&c);
    ListNode* l4 = vector2ListNode(&d);

    vector<ListNode*> lists;
    lists.push_back(l1);
    lists.push_back(l2);
    lists.push_back(l3);
    lists.push_back(l4);

    Solution obj;
    ListNode* result = obj.mergeKLists(lists);
    ListNode* ptr = result;

    while(ptr){
        cout<<ptr->val<<endl;
        ptr = ptr->next;
    }

    return 0;
}

File: _Linked List/23. Merge k Sorted Lists.py
from queue import PriorityQueue

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def mergeKLists(self, lists):
        # Merge list one by one

        if len(lists)==0 or len(lists)==1:
            return lists

        result = lists[0]

        for i in range(1,len(lists)):
            result = self.mergeTwoLists(result, lists[i])

        return result

    def mergeTwoLists(self, l1, l2):

        merge = ListNode(-1)
        ptr_merge = merge
        ptr1,ptr2 = l1,l2

        while ptr1 != None or ptr2 != None:

            x = ptr1.val if (ptr1 != None) else 9999999
            y = ptr2.val if (ptr2 != None) else 9999999

            ptr_merge.next = ListNode(min(x,y))
            ptr_merge = ptr_merge.next
            if x>y:
                ptr2 = ptr2.next
            else:
                ptr1 = ptr1.next

        return merge.next

    def mergeKLists1(self, lists):
        # Priority Queue

        head = point = ListNode(0)
        q = PriorityQueue()
        for l in lists:
            if l:
                q.put((l.val, l))
        while not q.empty():
            val, node = q.get()
            point.next = ListNode(val)
            point = point.next
            node = node.next
            if node:
                q.put((node.val, node))
        return head.next


if __name__ == "__main__":
    ls1 = [45, 47, 49, 80, 97]
    ls2 = [20, 24, 31, 65, 87]
    ls3 = [33, 45, 56, 64, 93]
    ls4 = [ 3, 19, 35, 38, 59]


    full_vector = [ls1,ls2,ls3,ls4]
    node_vector = []

    for vector in full_vector:
        linkedList = ListNode(0)
        ptr = linkedList

        for num in vector:
            ptr.next = ListNode(num)
            ptr = ptr.next

        node_vector.append(linkedList.next)

    obj = Solution()
    result = obj.mergeKLists1(lists=node_vector)

    ptr = result

    while ptr:
        print(ptr.val)
        ptr = ptr.next


    pass